File: jas/controller_test.py
from absl.testing import absltest

from controller import Controller
from document import Document
from element import Group, Layer, Line, Rect
from model import Model


class ControllerTest(absltest.TestCase):

    def test_default_document(self):
        ctrl = Controller()
        self.assertEqual(ctrl.document.title, "Untitled")
        self.assertEqual(len(ctrl.document.layers), 1)

    def test_initial_document(self):
        doc = Document(title="Test")
        model = Model(document=doc)
        ctrl = Controller(model=model)
        self.assertEqual(ctrl.document.title, "Test")

    def test_set_title(self):
        ctrl = Controller()
        ctrl.set_title("New Title")
        self.assertEqual(ctrl.document.title, "New Title")

    def test_add_layer(self):
        ctrl = Controller()
        layer = Layer(children=(Rect(x=0, y=0, width=10, height=10),), name="L1")
        ctrl.add_layer(layer)
        self.assertEqual(len(ctrl.document.layers), 2)
        self.assertEqual(ctrl.document.layers[1].name, "L1")

    def test_remove_layer(self):
        l1 = Layer(children=(), name="A")
        l2 = Layer(children=(), name="B")
        model = Model(document=Document(layers=(l1, l2)))
        ctrl = Controller(model=model)
        ctrl.remove_layer(0)
        self.assertEqual(len(ctrl.document.layers), 1)
        self.assertEqual(ctrl.document.layers[0].name, "B")

    def test_set_document(self):
        ctrl = Controller()
        new_doc = Document(title="Replaced")
        ctrl.set_document(new_doc)
        self.assertEqual(ctrl.document.title, "Replaced")

    def test_mutations_notify_model(self):
        model = Model()
        ctrl = Controller(model=model)
        received = []
        model.on_document_changed(lambda doc: received.append(doc.title))
        ctrl.set_title("Changed")
        self.assertEqual(received, ["Changed"])

    def test_model_immutability(self):
        ctrl = Controller()
        doc_before = ctrl.document
        ctrl.set_title("New")
        doc_after = ctrl.document
        self.assertEqual(doc_before.title, "Untitled")
        self.assertEqual(doc_after.title, "New")


class SelectionControllerTest(absltest.TestCase):

    def setUp(self):
        rect = Rect(x=0, y=0, width=10, height=10)
        line = Line(x1=0, y1=0, x2=5, y2=5)
        circle_line = Line(x1=1, y1=1, x2=2, y2=2)
        group = Group(children=(line, circle_line))
        layer = Layer(children=(rect, group), name="L0")
        doc = Document(layers=(layer,))
        self.model = Model(document=doc)
        self.ctrl = Controller(model=self.model)

    def test_set_selection(self):
        sel = frozenset({(0, 0)})
        self.ctrl.set_selection(sel)
        self.assertEqual(self.ctrl.document.selection, sel)

    def test_set_selection_clears(self):
        self.ctrl.set_selection(frozenset({(0, 0)}))
        self.ctrl.set_selection(frozenset())
        self.assertEqual(self.ctrl.document.selection, frozenset())

    def test_select_element_direct_child(self):
        """Clicking a direct child of a layer selects only that element."""
        self.ctrl.select_element((0, 0))
        self.assertEqual(self.ctrl.document.selection, frozenset({(0, 0)}))

    def test_select_element_in_group(self):
        """Clicking an element inside a Group selects all group children."""
        self.ctrl.select_element((0, 1, 0))
        self.assertEqual(
            self.ctrl.document.selection,
            frozenset({(0, 1, 0), (0, 1, 1)}),
        )

    def test_select_element_in_group_other_child(self):
        """Clicking a different child of the same Group selects the same set."""
        self.ctrl.select_element((0, 1, 1))
        self.assertEqual(
            self.ctrl.document.selection,
            frozenset({(0, 1, 0), (0, 1, 1)}),
        )

    def test_select_element_notifies_model(self):
        received = []
        self.model.on_document_changed(lambda doc: received.append(doc.selection))
        self.ctrl.select_element((0, 0))
        self.assertEqual(len(received), 1)
        self.assertEqual(received[0], frozenset({(0, 0)}))

    def test_select_element_layer_path(self):
        """Selecting a layer itself (path length 1) selects just the layer."""
        self.ctrl.select_element((0,))
        self.assertEqual(self.ctrl.document.selection, frozenset({(0,)}))

    def test_select_rect_hits_element(self):
        """Marquee covering the rect selects it."""
        self.ctrl.select_rect(-1, -1, 12, 12)
        self.assertIn((0, 0), self.ctrl.document.selection)

    def test_select_rect_misses_element(self):
        """Marquee outside all elements selects nothing."""
        self.ctrl.select_rect(100, 100, 10, 10)
        self.assertEqual(self.ctrl.document.selection, frozenset())

    def test_select_rect_group_expansion(self):
        """Marquee hitting one group child selects all group children."""
        # Layer has rect(0,0,10,10) and group(line(0,0→5,5), line(1,1→2,2)).
        # Use a rect that only hits the group lines (beyond rect's bounds).
        # line(0,0→5,5) has bounds (0,0,5,5), so a box at (10.5,0,5,5) misses
        # the rect but we need a box that hits only group children.
        # Actually, the rect and lines overlap at origin. Use a new setUp doc.
        rect_far = Rect(x=100, y=100, width=10, height=10)
        line1 = Line(x1=0, y1=0, x2=5, y2=5)
        line2 = Line(x1=1, y1=1, x2=2, y2=2)
        group = Group(children=(line1, line2))
        layer = Layer(children=(rect_far, group), name="L0")
        doc = Document(layers=(layer,))
        model = Model(document=doc)
        ctrl = Controller(model=model)
        # Marquee covers only the group lines, not the far rect
        ctrl.select_rect(-1, -1, 7, 7)
        self.assertEqual(
            ctrl.document.selection,
            frozenset({(0, 1, 0), (0, 1, 1)}),
        )

    def test_select_rect_replaces_previous(self):
        """Marquee selection replaces any prior selection."""
        self.ctrl.set_selection(frozenset({(0, 0)}))
        self.ctrl.select_rect(100, 100, 10, 10)
        self.assertEqual(self.ctrl.document.selection, frozenset())

    def test_select_rect_misses_diagonal_line_corner(self):
        """Marquee in the bounding box corner of a diagonal line should miss."""
        line = Line(x1=0, y1=0, x2=100, y2=100)
        layer = Layer(children=(line,), name="L0")
        doc = Document(layers=(layer,))
        ctrl = Controller(model=Model(document=doc))
        # Box in upper-right corner of bbox — far from the diagonal
        ctrl.select_rect(80, 0, 20, 20)
        self.assertEqual(ctrl.document.selection, frozenset())

    def test_select_rect_hits_diagonal_line(self):
        """Marquee crossing a diagonal line should select it."""
        line = Line(x1=0, y1=0, x2=100, y2=100)
        layer = Layer(children=(line,), name="L0")
        doc = Document(layers=(layer,))
        ctrl = Controller(model=Model(document=doc))
        # Box crossing the diagonal
        ctrl.select_rect(40, 40, 20, 20)
        self.assertIn((0, 0), ctrl.document.selection)

    def test_select_rect_stroke_only_rect_interior_misses(self):
        """Marquee inside a stroke-only rect should miss."""
        rect = Rect(x=0, y=0, width=100, height=100)
        layer = Layer(children=(rect,), name="L0")
        doc = Document(layers=(layer,))
        ctrl = Controller(model=Model(document=doc))
        ctrl.select_rect(30, 30, 10, 10)
        self.assertEqual(ctrl.document.selection, frozenset())

    def test_select_rect_filled_rect_interior_hits(self):
        """Marquee inside a filled rect should select it."""
        from element import Fill, Color
        rect = Rect(x=0, y=0, width=100, height=100,
                    fill=Fill(color=Color(1, 0, 0)))
        layer = Layer(children=(rect,), name="L0")
        doc = Document(layers=(layer,))
        ctrl = Controller(model=Model(document=doc))
        ctrl.select_rect(30, 30, 10, 10)
        self.assertIn((0, 0), ctrl.document.selection)

    def test_select_rect_multiple_elements(self):
        """Marquee covering both the rect and the group selects all."""
        self.ctrl.select_rect(-1, -1, 20, 20)
        self.assertIn((0, 0), self.ctrl.document.selection)
        self.assertIn((0, 1, 0), self.ctrl.document.selection)
        self.assertIn((0, 1, 1), self.ctrl.document.selection)


if __name__ == "__main__":
    absltest.main()

File: jas_ocaml/test/controller_test.ml
let () =
  let open Jas.Element in

  (* Test default document *)
  let ctrl = Jas.Controller.create () in
  assert (ctrl#document.Jas.Document.title = "Untitled");
  assert (List.length ctrl#document.Jas.Document.layers = 1);

  (* Test initial document *)
  let doc = Jas.Document.make_document ~title:"Test" [] in
  let model = Jas.Model.create ~document:doc () in
  let ctrl2 = Jas.Controller.create ~model () in
  assert (ctrl2#document.Jas.Document.title = "Test");

  (* Test set_title *)
  let ctrl3 = Jas.Controller.create () in
  ctrl3#set_title "New Title";
  assert (ctrl3#document.Jas.Document.title = "New Title");

  (* Test add_layer *)
  let ctrl4 = Jas.Controller.create () in
  let layer = make_layer ~name:"L1" [make_rect 0.0 0.0 10.0 10.0] in
  ctrl4#add_layer layer;
  assert (List.length ctrl4#document.Jas.Document.layers = 2);

  (* Test remove_layer *)
  let l1 = make_layer ~name:"A" [] in
  let l2 = make_layer ~name:"B" [] in
  let doc5 = Jas.Document.make_document [l1; l2] in
  let model5 = Jas.Model.create ~document:doc5 () in
  let ctrl5 = Jas.Controller.create ~model:model5 () in
  ctrl5#remove_layer 0;
  assert (List.length ctrl5#document.Jas.Document.layers = 1);
  (match List.hd ctrl5#document.Jas.Document.layers with
   | Layer { name; _ } -> assert (name = "B")
   | _ -> assert false);

  (* Test set_document *)
  let ctrl6 = Jas.Controller.create () in
  let new_doc = Jas.Document.make_document ~title:"Replaced" [] in
  ctrl6#set_document new_doc;
  assert (ctrl6#document.Jas.Document.title = "Replaced");

  (* Test mutations notify model *)
  let model7 = Jas.Model.create () in
  let ctrl7 = Jas.Controller.create ~model:model7 () in
  let received = ref [] in
  model7#on_document_changed (fun doc -> received := doc.Jas.Document.title :: !received);
  ctrl7#set_title "Changed";
  assert (!received = ["Changed"]);

  (* Test model immutability: old snapshots unchanged *)
  let ctrl8 = Jas.Controller.create () in
  let before = ctrl8#document in
  ctrl8#set_title "New";
  let after = ctrl8#document in
  assert (before.Jas.Document.title = "Untitled");
  assert (after.Jas.Document.title = "New");

  (* === Selection controller tests === *)

  let rect = make_rect 0.0 0.0 10.0 10.0 in
  let line1 = make_line 0.0 0.0 5.0 5.0 in
  let line2 = make_line 1.0 1.0 2.0 2.0 in
  let group = make_group [line1; line2] in
  let layer = make_layer ~name:"L0" [rect; group] in
  let doc_s = Jas.Document.make_document [layer] in
  let model_s = Jas.Model.create ~document:doc_s () in
  let ctrl_s = Jas.Controller.create ~model:model_s () in

  (* Test set_selection *)
  let sel = Jas.Document.PathSet.singleton [0; 0] in
  ctrl_s#set_selection sel;
  assert (Jas.Document.PathSet.equal ctrl_s#document.Jas.Document.selection sel);

  (* Test set_selection clears *)
  ctrl_s#set_selection Jas.Document.PathSet.empty;
  assert (Jas.Document.PathSet.is_empty ctrl_s#document.Jas.Document.selection);

  (* Test select_element: direct child of layer *)
  ctrl_s#select_element [0; 0];
  assert (Jas.Document.PathSet.equal ctrl_s#document.Jas.Document.selection
    (Jas.Document.PathSet.singleton [0; 0]));

  (* Test select_element: child inside a group selects all group children *)
  ctrl_s#select_element [0; 1; 0];
  let expected = Jas.Document.PathSet.of_list [[0; 1; 0]; [0; 1; 1]] in
  assert (Jas.Document.PathSet.equal ctrl_s#document.Jas.Document.selection expected);

  (* Test select_element: other child of same group *)
  ctrl_s#select_element [0; 1; 1];
  assert (Jas.Document.PathSet.equal ctrl_s#document.Jas.Document.selection expected);

  (* Test select_element: layer path *)
  ctrl_s#select_element [0];
  assert (Jas.Document.PathSet.equal ctrl_s#document.Jas.Document.selection
    (Jas.Document.PathSet.singleton [0]));

  (* Test select_element notifies model *)
  let model_n = Jas.Model.create ~document:doc_s () in
  let ctrl_n = Jas.Controller.create ~model:model_n () in
  let notify_count = ref 0 in
  model_n#on_document_changed (fun _ -> notify_count := !notify_count + 1);
  ctrl_n#select_element [0; 0];
  assert (!notify_count = 1);

  (* === select_rect tests === *)

  let rect_far = make_rect 100.0 100.0 10.0 10.0 in
  let sline1 = make_line 0.0 0.0 5.0 5.0 in
  let sline2 = make_line 1.0 1.0 2.0 2.0 in
  let sgroup = make_group [sline1; sline2] in
  let slayer = make_layer ~name:"L0" [rect_far; sgroup] in
  let sdoc = Jas.Document.make_document [slayer] in
  let smodel = Jas.Model.create ~document:sdoc () in
  let sctrl = Jas.Controller.create ~model:smodel () in

  (* select_rect hits element *)
  sctrl#select_rect 99.0 99.0 12.0 12.0;
  assert (Jas.Document.PathSet.mem [0; 0] sctrl#document.Jas.Document.selection);

  (* select_rect misses all *)
  sctrl#select_rect 200.0 200.0 10.0 10.0;
  assert (Jas.Document.PathSet.is_empty sctrl#document.Jas.Document.selection);

  (* select_rect group expansion *)
  sctrl#select_rect (-1.0) (-1.0) 7.0 7.0;
  let expected_sr = Jas.Document.PathSet.of_list [[0; 1; 0]; [0; 1; 1]] in
  assert (Jas.Document.PathSet.equal sctrl#document.Jas.Document.selection expected_sr);

  (* select_rect replaces previous *)
  sctrl#set_selection (Jas.Document.PathSet.singleton [0; 0]);
  sctrl#select_rect 200.0 200.0 10.0 10.0;
  assert (Jas.Document.PathSet.is_empty sctrl#document.Jas.Document.selection);

  (* select_rect multiple elements *)
  sctrl#select_rect (-1.0) (-1.0) 120.0 120.0;
  assert (Jas.Document.PathSet.mem [0; 0] sctrl#document.Jas.Document.selection);
  assert (Jas.Document.PathSet.mem [0; 1; 0] sctrl#document.Jas.Document.selection);
  assert (Jas.Document.PathSet.mem [0; 1; 1] sctrl#document.Jas.Document.selection);

  Printf.printf "All controller tests passed.\n"
